File: solutions/java/src/main/java/playground/algorithm/ValidParentheses.java
package playground.algorithm;

import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

public class ValidParentheses {

    public boolean isValid(String s) {
        Map<Character, Character> mapping = new HashMap<>();
        mapping.put(')', '(');
        mapping.put(']', '[');
        mapping.put('}', '{');

        Deque<Character> stack = new LinkedList<>();
        for (int i = 0; i < s.length(); i++) {
            Character ch = s.charAt(i);

            if (ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
                continue;
            }

            if (!stack.isEmpty() && stack.peek() == mapping.get(ch)) {
                stack.pop();
            } else {
                return false;
            }
        }

        return stack.isEmpty();
    }
}

File: solutions/go/src/playground/algorithm/valid_parentheses.go
package algorithm

func isValid(s string) bool {
	mapping := map[rune]rune{')': '(', ']': '[', '}': '{'}

	var stack []rune
	for _, c := range s {
		if c == '(' || c == '[' || c == '{' {
			stack = append(stack, c)
			continue
		}
		if len(stack) > 0 && stack[len(stack)-1] == mapping[c] {
			stack = stack[0 : len(stack)-1]
		} else {
			return false
		}
	}

	return len(stack) == 0
}
